app/classes/query/modules/ordering.rb

```diff
@@ -1,121 +1,219 @@
         add_join(:observations, :names)
         self.group = "images.id"
         "MIN(names.sort_name) ASC, images.when DESC"
-      elsif model == Location
+      end
+
+      def sort_locations_by_name
         if User.current_location_format == "scientific"
           "locations.scientific_name ASC"
         else
           "locations.name ASC"
         end
-      elsif model == LocationDescription
+      end
+
+      def sort_location_descriptions_by_name
         add_join(:locations)
         "locations.name ASC, location_descriptions.created_at ASC"
-      elsif model == Name
+      end
+
+      def sort_names_by_name
         "names.sort_name ASC"
-      elsif model == NameDescription
+      end
+
+      def sort_name_descriptions_by_name
         add_join(:names)
         "names.sort_name ASC, name_descriptions.created_at ASC"
-      elsif model == Observation
+      end
+
+      def sort_observations_by_name
         add_join(:names)
         "names.sort_name ASC, observations.when DESC"
-      elsif columns.include?("sort_name")
-        "#{table}.sort_name ASC"
-      elsif columns.include?("name")
-        "#{table}.name ASC"
-      elsif columns.include?("title")
-        "#{table}.title ASC"
       end
 
-    when "title", "login", "summary", "copyright_holder", "where",
-         "initial_det", "accession_number"
-      "#{table}.#{by} ASC" if columns.include?(by)
+      def sort_other_models_by_name(model)
+        if model.column_names.include?("sort_name")
+          "#{model.table_name}.sort_name ASC"
+        elsif model.column_names.include?("name")
+          "#{model.table_name}.name ASC"
+        elsif model.column_names.include?("title")
+          "#{model.table_name}.title ASC"
+        end
+      end
+
+      def sort_by_title(model)
+        "#{model.table_name}.title ASC" if model.column_names.include?("title")
+      end
+
+      def sort_by_login(model)
+        "#{model.table_name}.login ASC" if model.column_names.include?("login")
+      end
+
+      def sort_by_summary(model)
+        return unless model.column_names.include?("summary")
+
+        "#{model.table_name}.summary ASC"
+      end
+
+      def sort_by_copyright_holder(model)
+        return unless model.column_names.include?("copyright_holder")
+
+        "#{model.table_name}.copyright_holder ASC"
+      end
+
+      def sort_by_where(model)
+        "#{model.table_name}.where ASC" if model.column_names.include?("where")
+      end
+
+      def sort_by_initial_det(model)
+        return unless model.column_names.include?("initial_det")
+
+        "#{model.table_name}.initial_det ASC"
+      end
+
+      def sort_by_accession_number(model)
+        return unless model.column_names.include?("accession_number")
+
+        "#{model.table_name}.accession_number ASC"
+      end
+
+      def sort_by_user(model)
+        unless model.column_names.include?("user_id") || model == Herbarium
+          return
```

Comment:
@pellaea, @nimmolo 
This line is untested because:
The only relevant, sortable tables are `herbaria` or have `user_id` column
**Question Should I eliminate the condition? **

Reply:
Yes.

Why is there a special exception for Herbarium any way?  It looks like it serves no purpose.  Oh I get it, this condition *should never happen*, that's exactly why it's there.  We just wanted it to give a useful error message if we ever tried to sort a table without user_id by user... or apparently we wanted to remind ourselves never to try to sort herbaria by user... for whatever reason.  Removing this condition will just make it either 1. crash if there is no such column (still an error message!) or 2. actually sort herbaria by uiser... which is bad exactly why?  I think herbaria.user is just the one who created it.  I guess that's pretty meaningless.  But harmless.

I suspect this conversation has already consumed more resources than the exception would ever have saved!! :)  Get rid of it!